mysql/mysqldump.go

```diff
@@ -0,0 +1,578 @@
+	conv.SrcSchema[tableName] = schema.Table{
+		Name:        tableName,
+		ColNames:    colNames,
+		ColDefs:     colDef,
+		PrimaryKeys: keys}
+	for _, constraint := range stmt.Constraints {
+		processConstraint(conv, tableName, constraint, "CREATE TABLE")
+	}
+}
+
+func processConstraint(conv *internal.Conv, table string, constraint *ast.Constraint, s string) {
+	st := conv.SrcSchema[table]
+	switch ct := constraint.Tp; ct {
+	case ast.ConstraintPrimaryKey:
+		checkEmpty(conv, st.PrimaryKeys, s) // Drop any previous primary keys.
+		st.PrimaryKeys = toSchemaKeys(constraint.Keys)
+		// In Spanner, primary key columns are usually annotated with NOT NULL,
+		// but this can be omitted to allow NULL values in key columns.
+		// In MySQL, the primary key constraint is a combination of
+		// NOT NULL and UNIQUE i.e. primary keys must be NOT NULL.
+		// We preserve MySQL semantics and enforce NOT NULL.
+		updateCols(conv, ast.ConstraintPrimaryKey, constraint.Keys, st.ColDefs, table)
+	default:
+		updateCols(conv, ct, constraint.Keys, st.ColDefs, table)
+	}
+	conv.SrcSchema[table] = st
+}
+
+// toSchemaKeys converts a string list of MySQL primary keys to
+// schema primary keys.
+func toSchemaKeys(columns []*ast.IndexPartSpecification) (keys []schema.Key) {
+	for _, colname := range columns {
+		// MySQL primary keys have no notation of ascending/descending.
+		// We map them all into ascending primary keys.
+		keys = append(keys, schema.Key{Column: colname.Column.Name.String()})
+	}
+	return keys
+}
+
+func updateCols(conv *internal.Conv, ct ast.ConstraintType, colNames []*ast.IndexPartSpecification, colDef map[string]schema.Column, tableName string) {
+	for _, column := range colNames {
+		colName := column.Column.OrigColName()
+		cd := colDef[colName]
+		switch ct {
+		case ast.ConstraintForeignKey:
+			cd.Ignored.ForeignKey = true
+		case ast.ConstraintUniq:
+			cd.Unique = true
+		case ast.ConstraintCheck:
+			cd.Ignored.Check = true
+		case ast.ConstraintPrimaryKey:
+			cd.NotNull = true
+		default:
+			conv.Unexpected(fmt.Sprintf("Found constraint type of const value `%v` in table : %s , column : %s. Mapping of ConstraintType can be found here: 'https://godoc.org/github.com/pingcap/parser/ast#ConstraintType'", ct, tableName, colName))
+		}
+		colDef[colName] = cd
+	}
+}
+
+func processAlterTable(conv *internal.Conv, stmt *ast.AlterTableStmt) {
+	if stmt.Table == nil {
+		logStmtError(conv, stmt, fmt.Errorf("table is nil"))
+		return
+	}
+	tableName, err := getTableName(stmt.Table)
+	if err != nil {
+		logStmtError(conv, stmt, fmt.Errorf("can't get table name: %w", err))
+		return
+	}
+	if _, ok := conv.SrcSchema[tableName]; ok {
+		for _, item := range stmt.Specs {
+			switch alterType := item.Tp; alterType {
+			case ast.AlterTableAddConstraint:
+				processConstraint(conv, tableName, item.Constraint, "ALTER TABLE")
+				conv.SchemaStatement(reflect.TypeOf(stmt).String())
+			case ast.AlterTableModifyColumn:
+				var keys []schema.Key
```

Comment:
This block of code is using the variable 'keys' in an odd way.

Its declared at the top of the case block, then we append to it (but we can only append once) and then we use it to set ctable.PrimaryKeys.

Is AlterTableModifyColumn adding primary keys or replacing them? It's the latter isn't it?

If so, then suggest:
- dropping var declaration for keys
- changing the isPk block to:
				if isPk {
					ctable := conv.SrcSchema[tableName]
					ctable.PrimaryKeys = []schema.Key{{Column: colname}}
					conv.SrcSchema[tableName] = ctable
				}

Reply:
Also, we should probably call checkEmpty.

Reply:
Yes correct, we are replacing primary key. Ok, I will change it according to your suggestion. I recall it was done so that we do not need to do it inline but you are right, it's an odd way.
Sure, I will call checkEmpty() before overwriting.

Reply:
Done. Dropped var declaration ok `keys` and changed `isPk` block.
https://github.com/cloudspannerecosystem/harbourbridge/blob/99db369b2b826fc24ffbb760f3b3700cd0e04b0f/mysql/mysqldump.go#L266-L271